ee/katc/table.go

```diff
@@ -0,0 +1,178 @@
+	}
+	for i := 0; i < len(cfg.Columns); i += 1 {
+		columns = append(columns, table.ColumnDefinition{
+			Name: cfg.Columns[i],
+			Type: table.ColumnTypeText,
+		})
+		columnLookup[cfg.Columns[i]] = struct{}{}
+	}
+
+	return &katcTable{
+		cfg:          cfg,
+		columnLookup: columnLookup,
+		slogger: slogger.With(
+			"table_name", tableName,
+			"table_type", cfg.SourceType,
+			"table_source", cfg.Source,
+		),
+	}, columns
+}
+
+// generate handles queries against a KATC table.
+func (k *katcTable) generate(ctx context.Context, queryContext table.QueryContext) ([]map[string]string, error) {
+	// Fetch data from our table source
+	dataRaw, err := k.cfg.SourceType.dataFunc(ctx, k.slogger, k.cfg.Source, k.cfg.Query, getSourceConstraint(queryContext))
+	if err != nil {
+		return nil, fmt.Errorf("fetching data: %w", err)
+	}
+
+	// Process data
+	transformedResults := make([]map[string]string, 0)
+	for _, s := range dataRaw {
+		for _, dataRawRow := range s.rows {
+			// Make sure source is included in row data
+			rowData := map[string]string{
+				sourceColumnName: s.path,
+			}
+
+			// Run any needed transformations on the row data
+			for _, step := range k.cfg.RowTransformSteps {
+				dataRawRow, err = step.transformFunc(ctx, k.slogger, dataRawRow)
+				if err != nil {
+					return nil, fmt.Errorf("running transform func %s: %w", step.name, err)
+				}
+			}
+
+			// After transformations have been applied, we can cast the data from []byte
+			// to string to return to osquery.
+			for key, val := range dataRawRow {
+				rowData[key] = string(val)
+			}
+			transformedResults = append(transformedResults, rowData)
+		}
+	}
+
+	// Now, filter data to ensure we only return columns in k.columnLookup
+	filteredResults := make([]map[string]string, 0)
+	for _, row := range transformedResults {
+		filteredRow := make(map[string]string)
+		for column, data := range row {
+			if _, expectedColumn := k.columnLookup[column]; !expectedColumn {
+				// Silently discard the column+data
+				continue
+			}
+
+			filteredRow[column] = data
+		}
+
+		filteredResults = append(filteredResults, filteredRow)
+	}
+
+	return filteredResults, nil
+}
+
+// getSourceConstraint retrieves any constraints against the `source` column
+func getSourceConstraint(queryContext table.QueryContext) *table.ConstraintList {
+	sourceConstraint, sourceConstraintExists := queryContext.Constraints[sourceColumnName]
+	if sourceConstraintExists {
+		return &sourceConstraint
+	}
+	return nil
+}
+
+// checkSourceConstraints validates whether a given `source` matches the given constraints.
+func checkSourceConstraints(source string, sourceConstraints *table.ConstraintList) (bool, error) {
+	if sourceConstraints == nil {
+		return true, nil
+	}
+
+	validSource := true
+	for _, sourceConstraint := range sourceConstraints.Constraints {
+		switch sourceConstraint.Operator {
+		case table.OperatorEquals:
+			if source != sourceConstraint.Expression {
+				validSource = false
```

Comment:
why not just `return false, nil` in these spots instead of having the `validSource` variable?

Reply:
Nice catch, thank you! No good reason